sklearn/compose/_column_transformer.py

```diff
@@ -1371,3 +1467,115 @@ def __call__(self, df):
+        When False, the list never emits the warning nor updates
+        `warning_was_emitted``. This is used to obtain a quiet copy of the list
+        for use by the `ColumnTransformer` itself, so that the warning is only
+        shown when a user accesses it directly.
+    """
+
+    def __init__(
+        self,
+        columns,
+        *,
+        future_dtype=None,
+        warning_was_emitted=False,
+        warning_enabled=True,
+    ):
+        super().__init__(columns)
+        self.future_dtype = future_dtype
+        self.warning_was_emitted = warning_was_emitted
+        self.warning_enabled = warning_enabled
+
+    def __getitem__(self, index):
+        self._show_remainder_cols_warning()
+        return super().__getitem__(index)
+
+    def _show_remainder_cols_warning(self):
+        if self.warning_was_emitted or not self.warning_enabled:
+            return
+        self.warning_was_emitted = True
+        future_dtype_description = {
+            "str": "column names (of type str)",
+            "bool": "a mask array (of type bool)",
+            # shouldn't happen because we always initialize it with a
+            # non-default future_dtype
+            None: "a different type depending on the ColumnTransformer inputs",
+        }.get(self.future_dtype, self.future_dtype)
+        warnings.warn(
+            (
+                "The format of the columns of the 'remainder' transformer in"
+                " ColumnTransformer.transformers_ will change in a future version, to"
+                " match the format of the other transformers.\n  At the moment the"
+                " remainder columns are stored as indices (of type int).  With the same"
+                " ColumnTransformer configuration, in the future they will be stored as"
+                f" {future_dtype_description}.\n  To use the new behavior now and"
+                " suppress this warning, use"
+                " ColumnTransformer(force_int_remainder_cols=False).\n"
+            ),
+            category=FutureWarning,
+        )
+
+    def _repr_pretty_(self, printer, *_):
+        """Override display in ipython console, otherwise the class name is shown."""
+        printer.text(repr(self.data))
+
+
+def _with_dtype_warning_enabled_set_to(warning_enabled, transformers):
+    result = []
+    for name, trans, columns in transformers:
+        if isinstance(columns, _RemainderColsList):
+            columns = _RemainderColsList(
+                columns.data,
+                future_dtype=columns.future_dtype,
+                warning_was_emitted=columns.warning_was_emitted,
+                warning_enabled=warning_enabled,
+            )
+        result.append((name, trans, columns))
+    return result
+
+
+def _with_dtype_warnings_disabled(transformers):
+    """Obtain a shallow copy that does not emit warnings.
+
+    The remainder columns warning (if it exists) is disabled.
+    """
+    return _with_dtype_warning_enabled_set_to(False, transformers)
```

Comment:
I changed it to call this function directly. I had introduced the `enabled` and `disabled` wrappers because I thought it was a bit easier to figure out what they did without needing to check the `_RemaindersColsList` when reading the `ColumnTransformer` code